### PR TITLE
More Consul fixes

### PR DESCRIPTION
**Pull request by kyrias**

### CODE REVIEW

File: src/consul.rs [hunk 1]
```diff
@@ -429,51 +436,101 @@ impl ConsulSession {
...
+    // Consul documents that stale results are generally consistent within 50 ms, so let's sleep
+    // for that amount of time before we start checking to try to prevent spurious errors returned.
     //
-    // When the Consul server which is the Raft leader is restarted all KV reads by default return
-    // 500 errors until a new Raft leader is elected. For our usecase it's fine if the read value
-    // is a bit stale though.
-    peer_url.query_pairs_mut().append_pair("stale", "1");
+    // We still perform 5 retries at 1 second intervals since the maximum delay is theoretically
+    // unbounded.
+    tokio::time::sleep(Duration::from_millis(50)).await;
```
Comment (svenstaro) [inline src/consul.rs:454]: Seems somewhat arbitrary and hand-wavy by Consul but ok.

File: src/consul.rs [hunk 2]
```diff
@@ -429,51 +436,101 @@ impl ConsulSession {
...
+            break;
         }
+
+        trace!("Successfully fetched own node config from Consul");
     }
 }
 
+#[derive(Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct KVResponse {
```
Comment (svenstaro) [inline src/consul.rs:509]: Document what this is used for.

File: src/consul.rs [hunk 3]
```diff
@@ -490,5 +547,14 @@ async fn ensure_config_exists(
         index.replace(new_index);
     };
 
-    Ok(())
+    let res = res
+        .json::<Vec<KVResponse>>()
+        .await
+        .context("Failed to parse KV response")?;
+
+    ensure!(res.len() == 1, "Expected 1 KV response, got {}", res.len());
```
Comment (svenstaro) [inline src/consul.rs:555]: When would that be the case? Add a comment or improve the message to help someone debug this if this is important to enforce.

File: src/consul.rs [hunk 4]
```diff
@@ -490,5 +547,14 @@ async fn ensure_config_exists(
...
+
+    res[0]
+        .session
+        .ok_or_else(|| anyhow!("Key was not locked by any session"))
```
Comment (svenstaro) [inline src/consul.rs:559]: Would it be nicer to
```rust
res.first().expect("Expected blah").session...
```
?
Then we can be safe within the same expression as opposed to also requiring the `ensure!()`.

File: src/consul.rs [hunk 5]
```diff
@@ -429,51 +436,101 @@ impl ConsulSession {
...
+    let mut interval = interval(Duration::from_secs(1));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
     loop {
+        // Wait for an interval tick to have passed to prevent us from hammering the server with
+        // requests.
+        tokio::select! {
+            _ = config_token.cancelled() => {
+                trace!("Consul config handler was cancelled");
+                break;
+            },
```
Comment (svenstaro) [inline src/consul.rs:468]: Do we really need to check for cancelled here if we're already checking immediately again for this down below?

File: src/consul.rs [hunk 6]
```diff
@@ -429,51 +436,101 @@ impl ConsulSession {
...
             res = ensure_config_exists(&client, peer_url.clone(), &mut index) => res,
         };
-        if let Err(err) = res {
-            error!("Could not get own node config from Consul, cancelling: {err}");
+
+        let owner_id = match res {
```
Comment (svenstaro) [inline src/consul.rs:480]: Give this construct a quick comment on what it's trying to achieve.

**Review by svenstaro — CHANGES_REQUESTED** [review]
Good stuff! Some small nits.

File: src/consul.rs (hunk 4 above)
Comment (kyrias) [reply src/consul.rs]: :+1:

File: src/consul.rs (hunk 5 above)
Comment (svenstaro) [reply src/consul.rs:468]: Good point, we _just don't have the time_.

**Pull request merged**